bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/types/BUnionType.java

```diff
@@ -53,21 +64,219 @@ public BUnionType() {
+        this.mergeUnionType(unionType);
+    }
+
+    /**
+     * Constructor used when defining union type defs where cyclic reference is possible.
+     *
+     * @param unionType flags associated with the type
+     * @param typeName typename associated with the type
+     */
+    public BUnionType(BUnionType unionType, String typeName) {
+        super(typeName, unionType.pkg, unionType.valueClass);
+        this.readonly = unionType.readonly;
+        this.typeFlags = unionType.typeFlags;
+        this.memberTypes = new ArrayList<>(unionType.memberTypes.size());
+        this.mergeUnionType(unionType);
+    }
+
+    /**
+     * Constructor used when defining union type defs where cyclic reference is possible.
+     *
+     * @param name typename
+     * @param typeFlags flags associated with the type
+     * @param readonly boolean represents if the type is readonly
+     * @param isCyclic boolean represents if the type is cyclic
+     */
+    public BUnionType(String name, int typeFlags, boolean readonly, boolean isCyclic) {
+        super(name, null, Object.class);
+        this.typeFlags = typeFlags;
+        this.readonly = readonly;
+        this.memberTypes = null;
+        this.isCyclic = isCyclic;
+    }
+
+    public BUnionType(Type[] memberTypes, String name, int typeFlags, boolean readonly, boolean isCyclic) {
+        this(Arrays.asList(memberTypes), name, typeFlags, readonly, isCyclic);
+    }
+
+    public void setMemberTypes(Type[] members) {
+        if (members == null) {
+            return;
+        }
+        this.memberTypes = readonly ? getReadOnlyTypes(members) : Arrays.asList(members);
+        setFlagsBasedOnMembers();
+    }
+
+    public void setMemberTypes(List<Type> members) {
+        if (members == null) {
+            return;
+        }
+        if (members.isEmpty()) {
+            this.memberTypes = members;
+            return;
+        }
+        this.resolvingReadonly = true;
+        this.memberTypes = readonly ? getReadOnlyTypes(members, new HashSet<>(members.size())) : members;
+        this.resolvingReadonly = false;
+        setFlagsBasedOnMembers();
+    }
+
+    public void setMemberTypes(List<Type> members, Set<Type> unresolvedTypes) {
+        if (members == null) {
+            return;
+        }
+        if (members.isEmpty()) {
+            this.memberTypes = members;
+            return;
+        }
+        this.resolvingReadonly = true;
+        this.memberTypes = readonly ? getReadOnlyTypes(members, unresolvedTypes) : members;
+        this.resolvingReadonly = false;
+        setFlagsBasedOnMembers();
+    }
+
+    public void setCyclic(boolean isCyclic) {
+        this.isCyclic = isCyclic;
+    }
+
+    public boolean isNilable() {
+        if (memberTypes == null || memberTypes.isEmpty()) {
+            return true;
+        }
+
+        if (this.resolving) {
+            return false;
+        }
+
+        // TODO: use the flag
```

Comment:
Can we do this change now or should we create an issue?

Reply:
This is an existing TODO note. Let's create an issue.